File: android/src/main/java/guichaguri/trackplayer/player/components/FocusManager.java
package guichaguri.trackplayer.player.components;

import android.content.Context;
import android.media.AudioManager;
import android.media.AudioManager.OnAudioFocusChangeListener;

/**
 * @author Guilherme Chaguri
 */
public class FocusManager implements OnAudioFocusChangeListener {

    private final Context context;
    private boolean hasAudioFocus = false;

    public FocusManager(Context context) {
        this.context = context;
    }

    public boolean enable() {
        if(hasAudioFocus) return true;

        AudioManager manager = (AudioManager)context.getSystemService(Context.AUDIO_SERVICE);
        int r = manager.requestAudioFocus(this, AudioManager.STREAM_MUSIC, AudioManager.AUDIOFOCUS_GAIN);

        if(r == AudioManager.AUDIOFOCUS_REQUEST_GRANTED) {
            hasAudioFocus = true;
            return true;
        }
        return false;
    }

    public void disable() {
        if(!hasAudioFocus) return;

        AudioManager manager = (AudioManager)context.getSystemService(Context.AUDIO_SERVICE);
        manager.abandonAudioFocus(this);
        hasAudioFocus = false;
    }

    @Override
    public void onAudioFocusChange(int focus) {
        switch(focus) {
            case AudioManager.AUDIOFOCUS_LOSS:
            case AudioManager.AUDIOFOCUS_LOSS_TRANSIENT:
                // TODO pause everything
                break;
            case AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK:
                // TODO duck everything
                break;
            case AudioManager.AUDIOFOCUS_GAIN:
                // TODO play and "unduck" everything
                break;
        }
    }
}

File: android/src/main/java/com/doublesymmetry/kotlinaudio/players/components/FocusManager.kt
package com.doublesymmetry.kotlinaudio.players.components

import android.content.Context
import android.media.AudioManager
import android.util.Log
import androidx.core.content.ContextCompat
import androidx.media.AudioAttributesCompat
import androidx.media.AudioFocusRequestCompat
import androidx.media.AudioManagerCompat
import com.doublesymmetry.kotlinaudio.models.PlayerOptions

class FocusManager (
    private val context: Context,
    private val listener: AudioManager.OnAudioFocusChangeListener,
    private val options: PlayerOptions
) {
    var hasAudioFocus = false
    var alwaysPauseOnInterruption = options.alwaysPauseOnInterruption
    private var focus: AudioFocusRequestCompat? = null

    fun requestAudioFocus() {
        if (hasAudioFocus) return

        val manager = ContextCompat.getSystemService(context, AudioManager::class.java)

        focus = AudioFocusRequestCompat.Builder(AudioManagerCompat.AUDIOFOCUS_GAIN)
            .setOnAudioFocusChangeListener(listener)
            .setAudioAttributes(
                AudioAttributesCompat.Builder()
                    .setUsage(AudioAttributesCompat.USAGE_MEDIA)
                    .setContentType(AudioAttributesCompat.CONTENT_TYPE_MUSIC)
                    .build()
            )
            .setWillPauseWhenDucked(alwaysPauseOnInterruption)
            .build()

        val result: Int = if (manager != null && focus != null) {
            AudioManagerCompat.requestAudioFocus(manager, focus!!)
        } else {
            AudioManager.AUDIOFOCUS_REQUEST_FAILED
        }

        hasAudioFocus = (result == AudioManager.AUDIOFOCUS_REQUEST_GRANTED)
    }

    fun abandonAudioFocusIfHeld() {
        if (!hasAudioFocus) return

        val manager = ContextCompat.getSystemService(context, AudioManager::class.java)

        val result: Int = if (manager != null && focus != null) {
            AudioManagerCompat.abandonAudioFocusRequest(manager, focus!!)
        } else {
            AudioManager.AUDIOFOCUS_REQUEST_FAILED
        }

        hasAudioFocus = (result != AudioManager.AUDIOFOCUS_REQUEST_GRANTED)
    }

}